powerauth-push-server/src/main/java/io/getlime/push/controller/rest/PushDeviceController.java

```diff
@@ -93,17 +95,32 @@ public Response createDevice(@RequestBody ObjectRequest<CreateDeviceRequest> req
         }
         logger.info("Received createDevice request, app ID: {}, activation ID: {}, token: {}, platform: {}", requestObject.getAppId(),
                 requestObject.getActivationId(), maskPushToken(requestObject.getToken()), requestObject.getPlatform());
-        String errorMessage = CreateDeviceRequestValidator.validate(requestObject);
+        final String errorMessage = CreateDeviceRequestValidator.validate(requestObject);
         if (errorMessage != null) {
             throw new PushServerException(errorMessage);
         }
-        String appId = requestObject.getAppId();
-        String pushToken = requestObject.getToken();
-        String platform = requestObject.getPlatform();
-        String activationId = requestObject.getActivationId();
-        List<PushDeviceRegistrationEntity> devices = lookupDeviceRegistrations(appId, activationId, pushToken);
-        final AppCredentialsEntity appCredentials = findAppCredentials(appId);
-        PushDeviceRegistrationEntity device;
+
+        final AppCredentialsEntity appCredentials = findAppCredentials(requestObject.getAppId());
+
+        // In case of parallel requests to create new device, the createOrUpdateDevice may fail because of unique
+        // constraint violation. Try to repeat the save.
+        RetryTemplate.builder()
+                .retryOn(DataIntegrityViolationException.class)
+                .build()
```

Comment:
The default retry count looks good, but I agree with @romanstrobl , let's think about _backOffPolicy_.

```java
private volatile BackOffPolicy backOffPolicy = new NoBackOffPolicy();
private volatile RetryPolicy retryPolicy = new SimpleRetryPolicy(3);
```